node/node.go

```diff
@@ -516,61 +518,95 @@ func (node *Node) Start(ctx context.Context) error {
-
-	cctx, cancel := context.WithCancel(context.Background())
-	node.cancelSubscriptionsCtx = cancel
+	var syncCtx context.Context
+	syncCtx, node.cancelChainSync = context.WithCancel(context.Background())
 
-	go node.handleSubscription(cctx, node.processBlock, "processBlock", node.BlockSub, "BlockSub")
-	go node.handleSubscription(cctx, node.processMessage, "processMessage", node.MessageSub, "MessageSub")
-
-	node.HeaviestTipSetHandled = func() {}
-	node.HeaviestTipSetCh = node.ChainReader.HeadEvents().Sub(chain.NewHeadTopic)
+	// Wire up propagation of new chain heads from the chain store to other components.
 	head, err := node.PorcelainAPI.ChainHead()
 	if err != nil {
 		return errors.Wrap(err, "failed to get chain head")
 	}
-	go node.handleNewHeaviestTipSet(cctx, head)
+	go node.handleNewChainHeads(syncCtx, head)
 
 	if !node.OfflineMode {
+		// Start bootstrapper.
 		node.Bootstrapper.Start(context.Background())
-	}
 
-	if err := node.setupHeartbeatServices(ctx); err != nil {
-		return errors.Wrap(err, "failed to start heartbeat services")
+		// Establish a barrier to be released when the initial chain sync has completed.
+		// Services which depend on a more-or-less synced chain can wait for this before starting up.
+		chainSynced := moresync.NewLatch(1)
+
+		// Start up 'hello' handshake service
+		syncCallBack := func(pid peer.ID, cids []cid.Cid, height uint64) {
+			cidSet := types.NewTipSetKey(cids...)
+			err := node.Syncer.HandleNewTipset(context.Background(), cidSet, pid)
+			if err != nil {
+				log.Infof("error handling blocks: %s", cidSet.String())
+				return
+			}
+			// For now, consider the initial bootstrap done after the syncer has (synchronously)
+			// processed the chain up to the head reported by the first peer to respond to hello.
+			// This is an interim sequence until a secure network bootstrap is implemented:
+			// https://github.com/filecoin-project/go-filecoin/issues/2674.
+			// For now, we trust that the first node to respond will be a configured bootstrap node
+			// and that we trust that node to inform us of the chain head.
+			// TODO: when the syncer rejects too-far-ahead blocks received over pubsub, don't consider
+			// sync done until it's caught up enough that it will accept blocks from pubsub.
+			// This might require additional rounds of hello.
+			// See https://github.com/filecoin-project/go-filecoin/issues/1105
+			chainSynced.Done()
+		}
+		node.HelloSvc = hello.New(node.Host(), node.ChainReader.GenesisCid(), syncCallBack, node.PorcelainAPI.ChainHead, node.Repo.Config().Net, flags.Commit)
+
+		// Subscribe to block pubsub after the initial sync completes.
+		go func() {
+			chainSynced.Wait()
+			if syncCtx.Err() == nil {
+				// Subscribe to block pubsub topic to learn about new chain heads.
+				node.BlockSub, err = node.pubsubscribe(syncCtx, net.BlockTopic, node.processBlock)
+				if err != nil {
+					log.Errorf("failed to subscribe to %s: %s", net.MessageTopic, err)
```

Comment:
```suggestion
					log.Errorf("failed to subscribe to %s: %s", net.BlockTopic, err)
```

Reply:
Is there any point continuing if this operation fails? Part of me thinks the node should exit.

Reply:
Let's defer that to an operator for now. Longer term I think chain syncing should be runtime switchable on or off, independently of other node functions, so no failures should shut down a node.

Reply:
Alright, that sounds reasonable to me -- the above error message is still wrong (should state that we failed to subscribe to the BlockTopic) but we change that in a follow on.